internal/controller/nodestate_cleaner_controller.go

```diff
@@ -0,0 +1,133 @@
+// SPDX-License-Identifier:Apache-2.0
+
+package controller
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	frrk8sv1beta1 "github.com/metallb/frr-k8s/api/v1beta1"
+)
+
+const (
+	// PodNameAnnotation is the annotation key used to store the pod name in FRRNodeState.
+	PodNameAnnotation = "frrk8s.metallb.io/pod-name"
+)
+
+// NodeStateCleaner reconciles Pod objects to clean up FRRNodeState resources.
+type NodeStateCleaner struct {
+	client.Client
+	Scheme *runtime.Scheme
+	Logger log.Logger
+}
+
+// +kubebuilder:rbac:groups=frrk8s.metallb.io,resources=frrnodestates,verbs=get;list;watch;delete
+
+func (r *NodeStateCleaner) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	level.Info(r.Logger).Log("controller", "NodeStateCleaner", "start reconcile", req.String())
+	defer level.Info(r.Logger).Log("controller", "NodeStateCleaner", "end reconcile", req.String())
+
+	frrNodeStates := &frrk8sv1beta1.FRRNodeStateList{}
+	if err := r.Client.List(ctx, frrNodeStates); err != nil {
+		level.Error(r.Logger).Log("controller", "NodeStateCleaner", "failed to list FRRNodeStates", err)
+		return ctrl.Result{}, err
+	}
+
+	var errors []error
+	for _, nodeState := range frrNodeStates.Items {
+		podName, exists := nodeState.Annotations[PodNameAnnotation]
```

Comment:
we don't really need this, as we can match nodestate name to pod's node (fetching pods with `app.kubernetes.io/component: frr-k8s` label). unless you really like the explicitness of the relation between the two controllers, I think removing it and just matching states to nodes with pods is cleaner

Reply:
that was my first iteration, but I didn't like the idea of harvesting all the pods looking for the right one. Additionally, we use a different app annotation in charts vs manifests.
This is more efficient and straightforward imo.

Reply:
we can add add a new label to both if that's our problem, and efficiency/straightforwardness is debatable 😅 (2 controllers "interacting" through annotations vs listing all pods with a label)